### ISSUE TITLE
Old google-api-python-client versions are not PEP 440 compliant

### ISSUE THREAD

**Issue by lambda** [pypa/packaging#627]
We use PDM for managing dependencies of our Python projects; PDM in turn uses Packaging for parsing package versions.

One of our CI jobs which runs `pdm update` to check for package updates just failed when using the new Packaging 22.0 release, due to a failure to parse an older version of [google-api-python-client](https://pypi.org/project/google-api-python-client/), 1.0beta5prerelease.

```
pdm.termui:   Adding requirement google-api-python-client==2.13.0
Traceback (most recent call last):
  File "/usr/local/bin/pdm", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.9/site-packages/pdm/core.py", line 258, in main
    return Core().main(args)
  File "/usr/local/lib/python3.9/site-packages/pdm/core.py", line 190, in main
    raise cast(Exception, err).with_traceback(traceback)
  File "/usr/local/lib/python3.9/site-packages/pdm/core.py", line 185, in main
    f(project, options)
  File "/usr/local/lib/python3.9/site-packages/pdm/cli/commands/update.py", line 60, in handle
    actions.do_update(
  File "/usr/local/lib/python3.9/site-packages/pdm/cli/actions.py", line 384, in do_update
    resolved = do_lock(
  File "/usr/local/lib/python3.9/site-packages/pdm/cli/actions.py", line 95, in do_lock
    mapping, dependencies = resolve(
  File "/usr/local/lib/python3.9/site-packages/pdm/resolver/core.py", line 35, in resolve
    result = resolver.resolve(requirements, max_rounds)
  File "/usr/local/lib/python3.9/site-packages/resolvelib/resolvers.py", line 521, in resolve
    state = resolution.resolve(requirements, max_rounds=max_rounds)
  File "/usr/local/lib/python3.9/site-packages/resolvelib/resolvers.py", line 372, in resolve
    self._add_to_criteria(self.state.criteria, r, parent=None)
  File "/usr/local/lib/python3.9/site-packages/resolvelib/resolvers.py", line 172, in _add_to_criteria
    if not criterion.candidates:
  File "/usr/local/lib/python3.9/site-packages/resolvelib/structs.py", line 127, in __bool__
    next(iter(self))
  File "/usr/local/lib/python3.9/site-packages/pdm/resolver/providers.py", line 261, in matches_gen
    yield from super_find()
  File "/usr/local/lib/python3.9/site-packages/pdm/resolver/providers.py", line 150, in matches_gen
    candidates = self._find_candidates(reqs[0])
  File "/usr/local/lib/python3.9/site-packages/pdm/resolver/providers.py", line 132, in _find_candidates
    return self.repository.find_candidates(
  File "/usr/local/lib/python3.9/site-packages/pdm/models/repositories.py", line 150, in find_candidates
    cans = list(self._find_candidates(requirement))
  File "/usr/local/lib/python3.9/site-packages/pdm/models/repositories.py", line 338, in _find_candidates
    for c in finder.find_all_packages(requirement.project_name)
  File "/usr/local/lib/python3.9/site-packages/unearth/finder.py", line 241, in find_all_packages
    return sorted(
  File "/usr/local/lib/python3.9/site-packages/unearth/finder.py", line 192, in _sort_key
    parse_version(package.version) if package.version is not None else 0,
  File "/usr/local/lib/python3.9/site-packages/packaging/version.py", line 52, in parse
    return Version(version)
  File "/usr/local/lib/python3.9/site-packages/packaging/version.py", line 197, in __init__
    raise InvalidVersion(f"Invalid version: '{version}'")
packaging.version.InvalidVersion: Invalid version: '1.0beta5prerelease'
```

It looks like this was likely caused by caused by #407 , and we've implemented the workaround mentioned in #530 of pinning the packaging version to 21.3. #530 mentions 'For users who might be impacted by this, they have hopefully caught this issue in (a) their test suites, (b) through the "warnings" presented during the execution of legacy code, during the deprecation period, or (c) as part of warnings presented by pip.', but we are not using pip but rather PDM, and have not seen any such warnings or output.

**Comment by lambda** [comment]
Filed pdm-project/pdm#1554 since according to the deprecation mentioned in #530, it looks like it's likely a PDM bug that it doesn't handle this situation correctly.

**Comment by pradyunsg** [comment]
I'm not sure what's actionable here?

**Comment by pradyunsg** [comment]
For posterity, https://pypi.org/project/google-api-python-client/1.0beta5prerelease/ is from 2011 and over a decade old.

**Comment by uranusjr** [comment]
I don’t think we _want_ to do anything about this. This needs to be handled by pdm.

**Comment by lambda** [comment]
> I'm not sure what's actionable here, for `packaging`. Could you elaborate on what the motivation for filing this issue is?

I filed the issue here because it was a change to `packaging` that broke our builds, but as I investigated I realized that PDM was probably the culprit so I filed the ticket over there as well. Looks like there's a fix on the PDM side now (frostming/unearth#26).

I'm not sure there is anything specific to do here, so closing.
